File: streampark-common/src/main/scala/org/apache/streampark/common/conf/InternalConfigHolder.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.conf

import org.apache.streampark.common.constants.Constants
import org.apache.streampark.common.util.{Logger, SystemPropertyUtils}
import org.apache.streampark.common.util.Implicits._

import javax.annotation.{Nonnull, Nullable}

import java.util
import java.util.concurrent.ConcurrentHashMap

/**
 * Thread-safe configuration storage containers. All configurations will be automatically
 * initialized from the spring configuration items of the same name.
 */
object InternalConfigHolder extends Logger {

  private val initialCapacity = 45

  /** configuration values storage (key -> value) */
  private val confData = new ConcurrentHashMap[String, Any](initialCapacity)

  /** configuration key options storage (key -> ConfigOption) */
  private val confOptions =
    new ConcurrentHashMap[String, InternalOption](initialCapacity)

  /** Initialize the ConfigHub. */
  def initConfigHub(): Unit = {
    Seq(CommonConfig, K8sFlinkConfig)
  }

  /** Register the ConfigOption */
  private[conf] def register(@Nonnull conf: InternalOption): Unit = {
    confOptions.put(conf.key, conf)
    if (conf.defaultValue != null) {
      confData.put(conf.key, conf.defaultValue)
    }
  }

  /**
   * Get configuration value via ConfigOption.
   *
   * When using this api, the type must be explicitly declared and the relevant type will be
   * automatically converted to some extent. 1) in scala: val result: Long =
   * ConfigHub.get(K8sFlinkConfig.sglMetricTrackTaskTimeoutSec) 2) in java: Long result =
   * ConfigHub.get(K8sFlinkConfig.sglMetricTrackTaskTimeoutSec());
   *
   * @return
   *   return the defaultValue of ConfigOption when the value has not been overwritten.
   */
  @Nonnull
  def get[T](@Nonnull conf: InternalOption): T = {
    val value = confData.get(conf.key)
    if (value == null || value == conf.defaultValue) {
      val v = SystemPropertyUtils.get(conf.key)
      if (v != null) {
        if (v != value) {
          set(conf, v)
        }
        v.cast[T](conf.classType)
      } else {
        conf.defaultValue.asInstanceOf[T]
      }
    } else {
      value.toString.cast[T](conf.classType)
    }
  }

  /**
   * Get configuration value via key.
   *
   * When using this api, the type must be explicitly declared and the relevant type will be
   * automatically converted to some extent. 1) in scala: val result: Long =
   * ConfigHub.get("streampark.flink-k8s.tracking.polling-task-timeout-sec.cluster-metric") 2) in
   * java: Long result =
   * ConfigHub.get("streampark.flink-k8s.tracking.polling-task-timeout-sec.cluster-metric");
   *
   * @throws IllegalArgumentException
   *   when the key has not been registered to ConfigHub.
   * @return
   *   return the defaultValue of ConfigOption when the value has not been overwritten.
   */
  @throws[IllegalArgumentException]
  @Nonnull
  def get[T](@Nonnull key: String): T = {
    confData.get(key) match {
      case null =>
        confOptions.get(key) match {
          case null =>
            val config = getConfig(key)
            SystemPropertyUtils.get(key) match {
              case v if v != null => v.cast[T](config.classType)
              case _ =>
                throw new IllegalArgumentException(s"Config key has not been registered: $key")
            }
          case conf: InternalOption => conf.defaultValue.asInstanceOf[T]
        }
      case v => v.asInstanceOf[T]
    }
  }

  /**
   * Get registered ConfigOption by key.
   *
   * @return
   *   nullable
   */
  @Nullable
  def getConfig(key: String): InternalOption = {
    confOptions.get(key)
  }

  /** Get keys of all registered ConfigOption. */
  @Nonnull
  def keys(): JavaSet[String] = {
    val map = new util.HashMap[String, InternalOption](confOptions.size())
    map.putAll(confOptions)
    map.keySet()
  }

  /**
   * Overwritten configuration value.
   *
   * @param conf
   *   should not be null.
   * @param value
   *   the type of value should be same as conf.classType.
   * @throws IllegalArgumentException
   *   when the conf has not been registered, or the value type is not same as conf.classType.
   */
  @throws[IllegalArgumentException]
  def set(@Nonnull conf: InternalOption, value: Any): Unit = {
    if (!confOptions.containsKey(conf.key)) {
      throw new IllegalArgumentException(s"config key has not been registered: $conf")
    }
    value match {
      case null => confData.remove(conf.key)
      case v if conf.classType != v.getClass =>
        throw new IllegalArgumentException(
          s"config value type is not match of ${conf.key}, required: ${conf.classType}, actual: ${v.getClass}")
      case v =>
        SystemPropertyUtils.set(conf.key, v.toString)
        confData.put(conf.key, v)
    }
  }

  /** log the current configuration info. */
  def log(): Unit = {
    val configKeys = keys()
    logInfo(s"""Registered configs:
               |ConfigHub collected configs: ${configKeys.size}
               |  ${configKeys
                .map(key =>
                  s"$key = ${if (key.contains("password")) Constants.DEFAULT_DATAMASK_STRING
                    else get(key)}")
                .mkString("\n  ")}""".stripMargin)
  }

}

File: streampark-common/src/main/java/org/apache/streampark/common/conf/InternalConfigHolder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.common.conf;

import org.apache.streampark.common.constants.Constants;
import org.apache.streampark.common.util.StreamParkLoggerFactory;
import org.apache.streampark.common.util.StringCastUtils;
import org.apache.streampark.common.util.SystemPropertyUtils;

import org.apache.streampark.shaded.org.slf4j.Logger;

import javax.annotation.Nonnull;
import javax.annotation.Nullable;

import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

/**
 * Thread-safe configuration storage containers. All configurations will be automatically
 * initialized from the spring configuration items of the same name.
 */
public final class InternalConfigHolder {

    private static final Logger LOG =
        StreamParkLoggerFactory.loggerFactory().getLogger(InternalConfigHolder.class.getName());

    private static final int INITIAL_CAPACITY = 45;

    private static final Map<String, Object> CONF_DATA = new ConcurrentHashMap<>(INITIAL_CAPACITY);
    private static final Map<String, InternalOption> CONF_OPTIONS =
        new ConcurrentHashMap<>(INITIAL_CAPACITY);

    static {
        initConfigHub();
    }

    private InternalConfigHolder() {
    }

    /** Initialize the ConfigHub. */
    public static void initConfigHub() {
        CommonConfig.STREAMPARK_WORKSPACE_LOCAL();
        InternalOption unused = K8sFlinkConfig.jobStatusTrackTaskTimeoutSec;
    }

    static void register(@Nonnull InternalOption conf) {
        CONF_OPTIONS.put(conf.getKey(), conf);
        if (conf.getDefaultValue() != null) {
            CONF_DATA.put(conf.getKey(), conf.getDefaultValue());
        }
    }

    @Nonnull
    @SuppressWarnings("unchecked")
    public static <T> T get(@Nonnull InternalOption conf) {
        Object value = CONF_DATA.get(conf.getKey());
        if (value == null || value == conf.getDefaultValue()) {
            String v = SystemPropertyUtils.get(conf.getKey());
            if (v != null) {
                if (!v.equals(value)) {
                    set(conf, v);
                }
                return StringCastUtils.cast(v, conf.getClassType());
            } else {
                return (T) conf.getDefaultValue();
            }
        } else {
            return StringCastUtils.cast(value.toString(), conf.getClassType());
        }
    }

    @Nonnull
    @SuppressWarnings("unchecked")
    public static <T> T get(@Nonnull String key) {
        Object v = CONF_DATA.get(key);
        if (v == null) {
            InternalOption conf = CONF_OPTIONS.get(key);
            if (conf == null) {
                InternalOption config = getConfig(key);
                String sysVal = SystemPropertyUtils.get(key);
                if (sysVal != null) {
                    return StringCastUtils.cast(sysVal, config.getClassType());
                }
                throw new IllegalArgumentException("Config key has not been registered: " + key);
            }
            return (T) conf.getDefaultValue();
        }
        return (T) v;
    }

    @Nullable
    public static InternalOption getConfig(String key) {
        return CONF_OPTIONS.get(key);
    }

    @Nonnull
    public static Set<String> keys() {
        return CONF_OPTIONS.keySet();
    }

    public static void set(@Nonnull InternalOption conf, Object value) {
        if (!CONF_OPTIONS.containsKey(conf.getKey())) {
            throw new IllegalArgumentException("config key has not been registered: " + conf);
        }
        if (value == null) {
            CONF_DATA.remove(conf.getKey());
        } else if (conf.getClassType() != value.getClass()) {
            throw new IllegalArgumentException(
                "config value type is not match of "
                    + conf.getKey()
                    + ", required: "
                    + conf.getClassType()
                    + ", actual: "
                    + value.getClass());
        } else {
            SystemPropertyUtils.set(conf.getKey(), value.toString());
            CONF_DATA.put(conf.getKey(), value);
        }
    }

    public static void log() {
        Set<String> configKeys = keys();
        String details =
            configKeys.stream()
                .map(
                    key -> key
                        + " = "
                        + (key.contains("password")
                            ? Constants.DEFAULT_DATAMASK_STRING
                            : get(key)))
                .collect(Collectors.joining("\n  "));
        LOG.info(
            "[StreamPark] Registered configs:\nConfigHub collected configs: {}\n  {}",
            configKeys.size(),
            details);
    }
}
